// @starknet-react/chains@3.1.0/dist/index.js
// src/devnet.ts
var devnet = {
  id: BigInt("0x534e5f5345504f4c4941"),
  network: "devnet",
  name: "Starknet Devnet",
  nativeCurrency: {
    address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
    name: "Ether",
    symbol: "ETH",
    decimals: 18
  },
  testnet: true,
  rpcUrls: {
    default: {
      http: []
    },
    public: {
      http: ["http://localhost:5050/rpc"]
    }
  }
};

// src/starknet.ts
var mainnet = {
  id: BigInt("0x534e5f4d41494e"),
  network: "mainnet",
  name: "Starknet",
  nativeCurrency: {
    address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
    name: "Ether",
    symbol: "ETH",
    decimals: 18
  },
  rpcUrls: {
    alchemy: {
      http: ["https://starknet-mainnet.g.alchemy.com/v2"]
    },
    blast: {
      http: ["https://starknet-mainnet.blastapi.io"]
    },
    infura: {
      http: ["https://starknet-mainnet.infura.io/v3"]
    },
    lava: {
      http: ["https://g.w.lavanet.xyz:443/gateway/strk/rpc-http"]
    },
    nethermind: {
      http: ["https://rpc.nethermind.io/mainnet-juno"]
    },
    reddio: {
      http: ["https://starknet-mainnet.reddio.com"]
    },
    cartridge: {
      http: ["https://api.cartridge.gg/x/starknet/mainnet"]
    },
    default: {
      http: []
    },
    public: {
      http: [
        "https://starknet-mainnet.public.blastapi.io/rpc/v0_7",
        "https://rpc.starknet.lava.build",
        "https://free-rpc.nethermind.io/mainnet-juno/v0_7"
      ]
    }
  },
  explorers: {
    starkCompass: ["https://www.starkcompass.com"],
    starkscan: ["https://starkscan.co"],
    viewblock: ["https://viewblock.io/starknet"],
    voyager: ["https://voyager.online"]
  }
};
var sepolia = {
  id: BigInt("0x534e5f5345504f4c4941"),
  network: "sepolia",
  name: "Starknet Sepolia Testnet",
  nativeCurrency: {
    address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
    name: "Ether",
    symbol: "ETH",
    decimals: 18
  },
  testnet: true,
  rpcUrls: {
    blast: {
      http: ["https://starknet-sepolia.blastapi.io"]
    },
    infura: {
      http: ["https://starknet-sepolia.infura.io/v3"]
    },
    nethermind: {
      http: ["https://rpc.nethermind.io/sepolia-juno"]
    },
    reddio: {
      http: ["https://starknet-sepolia.reddio.com"]
    },
    cartridge: {
      http: ["https://api.cartridge.gg/x/starknet/sepolia"]
    },
    default: {
      http: []
    },
    public: {
      http: [
        "https://starknet-sepolia.public.blastapi.io",
        "https://free-rpc.nethermind.io/sepolia-juno"
      ]
    }
  },
  explorers: {
    starkscan: ["https://sepolia.starkscan.co"],
    voyager: ["https://sepolia.voyager.online"]
  }
};

// src/slot.ts
function getSlotChain(projectId) {
  return {
    id: BigInt(projectId),
    network: `slot-${projectId}`,
    name: `${projectId}`,
    nativeCurrency: {
      address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
      name: "Ether",
      symbol: "ETH",
      decimals: 18
    },
    testnet: true,
    rpcUrls: {
      default: {
        http: []
      },
      public: {
        http: [`https://api.cartridge.gg/x/${projectId}/katana`]
      }
    }
  };
}
export {
  devnet,
  getSlotChain,
  mainnet,
  sepolia
};
