// @dynamic-labs/sdk-react-core@4.0.0-alpha.38/src/lib/shared/assets/currency.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

var _rect, _path, _path2, _ellipse;
var _excluded = ["title", "titleId"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) { ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) { o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) { if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } } return t; }
var SvgCurrency = function SvgCurrency(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React__namespace.createElement("svg", _extends({
    width: 28,
    height: 28,
    viewBox: "0 0 28 28",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React__namespace.createElement("title", {
    id: titleId
  }, title) : null, _rect || (_rect = /*#__PURE__*/React__namespace.createElement("rect", {
    width: 28,
    height: 28,
    rx: 14,
    fill: "#4779FF"
  })), _path || (_path = /*#__PURE__*/React__namespace.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M20.252 17.036C18.976 18.398 16.654 19.31 14 19.31c-2.65 0-4.97-.909-6.246-2.268-.004.646-.004 1.406-.004 2.293 0 1.726 2.798 3.125 6.25 3.125s6.25-1.399 6.25-3.125c0-.192.002-.406.004-.637.005-.496.01-1.069-.002-1.662Z",
    fill: "#fff"
  })), _path2 || (_path2 = /*#__PURE__*/React__namespace.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M20.23 11.85c-1.28 1.35-3.591 2.252-6.23 2.252-2.65 0-4.969-.909-6.245-2.267-.005.64-.005 1.398-.005 2.292 0 1.726 2.798 3.125 6.25 3.125s6.25-1.399 6.25-3.125c0-.959 0-1.703-.02-2.276Z",
    fill: "#fff"
  })), _ellipse || (_ellipse = /*#__PURE__*/React__namespace.createElement("ellipse", {
    cx: 14,
    cy: 8.893,
    rx: 6.25,
    ry: 3.125,
    fill: "#fff"
  })));
};

exports.ReactComponent = SvgCurrency;
