// @dynamic-labs/starknet@4.0.0-alpha.38/src/wallets/argent/webwallet/webwallet.js
'use client'
import { __awaiter } from '../../../../_virtual/_tslib.js';
import { connect } from 'starknetkit';
import { WebWalletConnector } from 'starknetkit/webwallet';
import { ArgentBase } from '../base.js';

class WebWallet extends ArgentBase {
    constructor(opts) {
        super('Argent X - Web', 'argentWebWallet', opts);
        this.overrideKey = 'argentxweb';
        this.canConnectViaCustodialService = true;
    }
    isInstalledOnBrowser() {
        return false;
    }
    prompt(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { connectorData, wallet } = yield connect({
                connectors: [new WebWalletConnector({ url: 'https://web.argent.xyz' })],
                modalMode: options.silently ? 'neverAsk' : 'canAsk',
            });
            return {
                data: {
                    account: connectorData === null || connectorData === void 0 ? void 0 : connectorData.account,
                    chainId: connectorData === null || connectorData === void 0 ? void 0 : connectorData.chainId,
                },
                wallet: wallet !== null && wallet !== void 0 ? wallet : undefined,
            };
        });
    }
}

export { WebWallet };
