// @dynamic-labs/starknet@4.0.0-alpha.38/src/wallets/argent/injected/argentx.js
'use client'
import { __awaiter } from '../../../../_virtual/_tslib.js';
import { connect } from 'starknetkit';
import { InjectedConnector } from 'starknetkit/injected';
import { ArgentBase } from '../base.js';

class ArgentX extends ArgentBase {
    constructor(opts) {
        super('Argent X', 'argentX', opts);
        this.overrideKey = 'argentx';
    }
    prompt() {
        return __awaiter(this, void 0, void 0, function* () {
            const { connectorData, wallet } = yield connect({
                connectors: [new InjectedConnector({ options: { id: this.id } })],
                modalMode: 'canAsk',
            });
            return {
                data: {
                    account: connectorData === null || connectorData === void 0 ? void 0 : connectorData.account,
                    chainId: connectorData === null || connectorData === void 0 ? void 0 : connectorData.chainId,
                },
                wallet: wallet !== null && wallet !== void 0 ? wallet : undefined,
            };
        });
    }
}

export { ArgentX };
