// starknetkit@2.3.3/dist/index-fd75fb40.js
var N = Object.defineProperty;
var k = (e, n, t) => n in e ? N(e, n, { enumerable: !0, configurable: !0, writable: !0, value: t }) : e[n] = t;
var s = (e, n, t) => (k(e, typeof n != "symbol" ? n + "" : n, t), t);
import { Account as v, constants as w } from "starknet";
import { C as E, P, a as o, b as h, U as u, r as M } from "./connector-c50f5de7.js";
import { g as R } from "./getStarknetChainId-7c4b3163.js";
import { I as S } from "./index-d6560940.js";
const L = {
  mainnet: "https://starknet-mainnet.public.blastapi.io",
  testnet: "https://starknet-sepolia.public.blastapi.io"
}, O = {
  mainnet: "https://rpc.starknet.lava.build",
  testnet: "https://rpc.starknet-sepolia.lava.build"
}, d = [L, O];
function q() {
  const e = Math.floor(Math.random() * d.length);
  return d[e];
}
const B = () => Object.keys(localStorage).some(
  (e) => e === "walletconnect" || e.startsWith("wc@2:")
), p = () => {
  if (B()) {
    delete localStorage.walletconnect;
    for (const e in localStorage)
      e.startsWith("wc@2:") && delete localStorage[e];
  }
}, _ = "f2e613881f7a0e811295cdd57999e31b", g = `<svg
    width="32"
    height="32"
    viewBox="0 0 32 32"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect width="32" height="32" rx="8" fill="#FF875B" />
    <path
      d="M18.316 8H13.684C13.5292 8 13.4052 8.1272 13.4018 8.28531C13.3082 12.7296 11.0323 16.9477 7.11513 19.9355C6.99077 20.0303 6.96243 20.2085 7.05335 20.3369L9.76349 24.1654C9.85569 24.2957 10.0353 24.3251 10.1618 24.2294C12.6111 22.3734 14.5812 20.1345 16 17.6529C17.4187 20.1345 19.389 22.3734 21.8383 24.2294C21.9646 24.3251 22.1443 24.2957 22.2366 24.1654L24.9467 20.3369C25.0375 20.2085 25.0092 20.0303 24.885 19.9355C20.9676 16.9477 18.6918 12.7296 18.5983 8.28531C18.5949 8.1272 18.4708 8 18.316 8Z"
      fill="white"
    />
  </svg>`, U = () => typeof window > "u" || !window?.starknet_argentX ? !1 : window?.starknet_argentX?.isInAppBrowser;
class W extends E {
  constructor(t) {
    super();
    s(this, "_wallet", null);
    s(this, "_options");
    this._options = t;
  }
  available() {
    return !0;
  }
  async ready() {
    if (!this._wallet)
      return !1;
    try {
      return (await this._wallet.request({
        type: "wallet_getPermissions"
      })).includes(P.ACCOUNTS);
    } catch {
      return !1;
    }
  }
  get id() {
    return "argentMobile";
  }
  get name() {
    return "Argent (mobile)";
  }
  get icon() {
    return {
      dark: g,
      light: g
    };
  }
  get wallet() {
    if (!this._wallet)
      throw new o();
    return this._wallet;
  }
  async connect() {
    if (await this.ensureWallet(), !this._wallet)
      throw new h();
    const t = await this._wallet.request({
      type: "wallet_requestAccounts",
      params: { silent_mode: !1 }
      // explicit to show the modal
    }), a = await this.chainId();
    return {
      account: t[0],
      chainId: a
    };
  }
  async disconnect() {
    if (await this._wallet.disable(), p(), !this.available() && !this._wallet)
      throw new h();
    this._wallet = null;
  }
  async account(t) {
    if (!this._wallet)
      throw new o();
    const a = await this._wallet.request({
      type: "wallet_requestAccounts",
      params: { silent_mode: !0 }
    });
    return new v(t, a[0], "");
  }
  async chainId() {
    if (!this._wallet)
      throw new o();
    const t = await this._wallet.request({
      type: "wallet_requestChainId"
    }), a = R(t);
    return BigInt(a);
  }
  async request(t) {
    if (!this._wallet)
      throw new o();
    try {
      return await this._wallet.request(t);
    } catch {
      throw new u();
    }
  }
  // needed, methods required by starknet-react. Otherwise an exception is throwd
  async initEventListener(t) {
    if (!this._wallet)
      throw new o();
    this._wallet.on("accountsChanged", t);
  }
  // needed, methods required by starknet-react. Otherwise an exception is throwd
  async removeEventListener(t) {
    if (!this._wallet)
      throw new o();
    this._wallet.off("accountsChanged", t), this._wallet = null;
  }
  async ensureWallet() {
    const { getStarknetWindowObject: t } = await import("./index-233476cf.js"), { chainId: a, projectId: r, dappName: f, description: m, url: C, icons: I, rpcUrl: b } = this._options, l = q(), y = b ?? (!a || a === w.NetworkName.SN_MAIN ? l.mainnet : l.testnet), A = {
      chainId: a ?? w.NetworkName.SN_MAIN,
      name: f,
      projectId: r ?? _,
      description: m,
      url: C,
      icons: I,
      rpcUrl: y
    };
    r === _ && (console.log("========= NOTICE ========="), console.log(
      "While your application will continue to function, we highly recommended"
    ), console.log("signing up for your own API keys."), console.log(
      "Go to WalletConnect Cloud (https://cloud.walletconnect.com) and create a new account."
    ), console.log(
      "Once your account is created, create a new project and collect the Project ID"
    ), console.log("=========================="));
    const i = await t(A);
    if (!i)
      throw new u();
    this._wallet = i;
    const c = this._wallet;
    await c.enable(), c.client.on("session_delete", () => {
      p(), this._wallet = null, M(), document.dispatchEvent(new Event("wallet_disconnected"));
    });
  }
}
class X {
  static init({
    options: n,
    inAppBrowserOptions: t
  }) {
    return U() ? new S({
      options: { id: "argentX", ...t }
    }) : new W(n);
  }
}
export {
  W as A,
  X as a,
  U as i,
  p as r
};
